src/NuGet.Core/NuGet.Commands/SourcesCommands/SourcesRunner.cs

```diff
@@ -0,0 +1,299 @@
+                                Strings.SourcesCommandInvalidSource));
+                }
+
+                // If the user is updating the source, verify we don't have a duplicate.
+                var duplicateSource = args.SourceProvider.GetPackageSourceBySource(args.Source);
+                if (duplicateSource != null)
+                {
+                    args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandUniqueSource));
+                }
+
+                existingSource = new Configuration.PackageSource(args.Source, existingSource.Name);
+            }
+
+            ValidateCredentials(args);
+
+            if (!string.IsNullOrEmpty(args.Username))
+            {
+                var hasExistingAuthTypes = existingSource.Credentials?.ValidAuthenticationTypes.Any() ?? false;
+                if (hasExistingAuthTypes && string.IsNullOrEmpty(args.ValidAuthenticationTypes))
+                {
+                    args.LogQuiet(string.Format(CultureInfo.CurrentCulture,
+                        Strings.SourcesCommandClearingExistingAuthTypes, args.Name));
+                }
+
+                var credentials = Configuration.PackageSourceCredential.FromUserInput(
+                    args.Name,
+                    args.Username,
+                    args.Password,
+                    args.StorePasswordInClearText,
+                    args.ValidAuthenticationTypes);
+                existingSource.Credentials = credentials;
+            }
+
+            args.SourceProvider.UpdatePackageSource(existingSource, updateCredentials: existingSource.Credentials != null, updateEnabled: false);
+
+            args.LogMinimal(string.Format(CultureInfo.CurrentCulture,
+                    Strings.SourcesCommandUpdateSuccessful, args.Name));
+        }
+
+        private static void ValidateCredentials(SourcesArgs args)
+        {
+            var isUsernameEmpty = string.IsNullOrEmpty(args.Username);
+            var isPasswordEmpty = string.IsNullOrEmpty(args.Password);
+            var isAuthTypesEmpty = string.IsNullOrEmpty(args.ValidAuthenticationTypes);
+
+            if (isUsernameEmpty ^ isPasswordEmpty)
+            {
+                // If only one of them is set, throw.
+                args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandCredentialsRequired));
+            }
+
+            if (isPasswordEmpty && !isAuthTypesEmpty)
+            {
+                // can't specify auth types without credentials
+                args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandCredentialsRequiredWithAuthTypes));
+            }
+        }
+
+        private static void PrintRegisteredSourcesShort(SourcesArgs args)
+        {
+            var sourcesList = args.SourceProvider.LoadPackageSources().ToList();
```

Comment:
Invoking ToList() is not necessary as we don't need a cached copy of results apart iterating sourcesList in a for-each loop.

Reply:
boilerplate reply: the code in SourcesRunner.cs was moved from SourcesCommand.cs - I have tried to make the minimal changes to it. That said, I'll review each of these comments. thx.

Reply:
fixed in next commit.